### PR TITLE
Refactor DeviceScopes out into a separate structure.

### PR DESCRIPTION
**Pull request by bgogul**

### CODE REVIEW

File: Sources/TensorFlow/Core/Runtime.swift [hunk 1]
```diff
@@ -1281,6 +1264,32 @@ class _ThreadLocalState {
...
+    }
+
+    @usableFromInline
+    func popDevice() {
+        internalConsistencyCheck(deviceStack.popLast() != nil)
+    }
+}
+
 @usableFromInline
 @_cdecl("_swift_tfc_OpSetDeviceFromScope")
```
Comment (rxwei) [inline Sources/TensorFlow/Core/Runtime.swift]: I think this can be deleted now.

File: Sources/TensorFlow/Core/Runtime.swift [hunk 2]
```diff
@@ -1281,6 +1264,32 @@ class _ThreadLocalState {
...
 
+/// Stack of devices that models nested calls to withDevice/withDefaultDevice. Devices are
+/// represented by their names in TensorFlow notation. See documentation for
+/// `withDevice(named:perform:)` to learn about device names.
+///
+/// All TensorFlow operations will be put on the topmost device on the stack. When the stack is
+/// empty or the topmost device is `nil`, that allows TensorFlow to place operations on any device
+/// that it sees fit.
+@usableFromInline
+class DeviceScopes {
```
Comment (pschuh) [inline Sources/TensorFlow/Core/Runtime.swift]: Why class? is there a reason for multiple threadlocalstate to share the same devicestack?

Comment (bgogul) [reply Sources/TensorFlow/Core/Runtime.swift]: Doesn't have to be. Changed `class` -> `struct` and made the related changes.

File: Sources/TensorFlow/Core/Runtime.swift (hunk 1 above)
Comment (bgogul) [reply Sources/TensorFlow/Core/Runtime.swift]: This is still used in EagerExecution.swift. However, those use cases can be rewritten. Will send out a separate PR for that.

Comment (bgogul) [reply Sources/TensorFlow/Core/Runtime.swift]: It looks like you were only referring to `_cdecl`. Will send out a PR cleaning this up and a few other uses.

Comment (rxwei) [reply Sources/TensorFlow/Core/Runtime.swift]: Yeah it'd be nice to have EagerExecution.swift use C API directly.

**Pull request merged**